tensorizer/serialization.py

```diff
@@ -1674,6 +1710,31 @@ def __init__(
             )
             if not self._metadata:
                 raise ValueError("Tensor index in the file is empty")
+
+            self._headers: Optional[
+                Dict[_TensorPath, _TensorHeaderDeserializer]
+            ]
+            if version_number >= HEADERS_AT_TOP_TENSORIZER_VERSION:
+                metadata_ordered = sorted(
+                    self._metadata.values(), key=operator.attrgetter("offset")
+                )
+                self._headers = {}
+                for entry in metadata_ordered:
+                    if self._file.tell() > entry.offset:
+                        raise ValueError("Header offsets overlap or are wrong")
+                    self._file.seek(entry.offset)
+                    header = _TensorHeaderDeserializer.from_io(
+                        version_number,
+                        self._file,
+                        zero_hashes=True,
+                        check_crypt_info=self._has_crypt_info,
+                    )
+                    if header is None:
+                        raise KeyError("Unexpected empty header")
+                    self._headers[entry.name] = header
```

Comment:
We should probably read the entire metadata section including these headers in one or two `read` (or `readv`) operations and just use the `_TensorHeaderDeserializer` buffer constructor instead of `from_io`.

Reply:
Agreed!

Reply:
https://github.com/coreweave/tensorizer/issues/148